### PR TITLE
feat: geolocation setter in sendgrid-ruby for GDPR compliance

### PR DESCRIPTION
**Pull request by manisha1997**
feat!: Geolocation setter for sendgrid ruby
Ticket: [DII-1232](https://issues.corp.twilio.com/browse/DII-1232)

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guidelines](https://github.com/sendgrid/sendgrid-ruby/blob/main/CONTRIBUTING.md) and my PR follows them
- [x] I have titled the PR appropriately
- [x] I have updated my branch with the main branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added the necessary documentation about the functionality in the appropriate .md file
- [x] I have added inline documentation to the code I modified

### CODE REVIEW

**Review by tiwarishubham635 — COMMENTED** [review]
LGTM!

File: lib/sendgrid/base_interface.rb [hunk 1]
```diff
@@ -16,9 +16,15 @@ class BaseInterface
...
-  def initialize(auth:, host:, request_headers: nil, version: nil, impersonate_subuser: nil, http_options: {})
+  def initialize(auth:, host: nil, region: nil, request_headers: nil, version: nil, impersonate_subuser: nil, http_options: {})
     @auth = auth
-    @host = host
+    if host.nil? && region.nil?
+      @host = "https://api.sendgrid.com"
+    elsif !region.nil?
+      data_residency(region: region)
+    else
+      update_host(host: host)
```
Comment (shrutiburman) [inline lib/sendgrid/base_interface.rb:26]: Why is this method required? If host is provided, it'll anyways get set in line # 42.

File: lib/sendgrid/base_interface.rb [hunk 2]
```diff
@@ -16,9 +16,15 @@ class BaseInterface
   #                              in the "On-Behalf-Of" header
   #   - +http_options+ -> http options that you want to be globally applied to each request
   #
-  def initialize(auth:, host:, request_headers: nil, version: nil, impersonate_subuser: nil, http_options: {})
+  def initialize(auth:, host: nil, region: nil, request_headers: nil, version: nil, impersonate_subuser: nil, http_options: {})
```
[the remark this answers is not in the document]
Comment (manisha1997) [reply lib/sendgrid/base_interface.rb]: fixed, that nil was unnecesssary

File: lib/sendgrid/base_interface.rb (hunk 1 above)
Comment (manisha1997) [reply lib/sendgrid/base_interface.rb]: fixed this

File: lib/sendgrid/base_interface.rb [hunk 3]
```diff
@@ -16,9 +16,15 @@ class BaseInterface
   #                              in the "On-Behalf-Of" header
   #   - +http_options+ -> http options that you want to be globally applied to each request
   #
-  def initialize(auth:, host:, request_headers: nil, version: nil, impersonate_subuser: nil, http_options: {})
+  def initialize(auth:, host:, region: nil, request_headers: nil, version: nil, impersonate_subuser: nil, http_options: {})
     @auth = auth
-    @host = host
+    if host.nil? && region.nil?
```
Comment (shrutiburman) [inline lib/sendgrid/base_interface.rb:21]: If host is not default to `nil`, can it be `nil` now? 
How about this? 

```
if !region.nil?
      data_residency(region: region)
    else
      @host = host
```

Comment (manisha1997) [reply lib/sendgrid/base_interface.rb]: adressed

**Review by tiwarishubham635 — APPROVED** [review]
LGTM

**Pull request merged**